### DISCUSSION TITLE
Useless explicit @db.ObjectId

### DISCUSSION THREAD

**Discussion by jasonkuhrt** [prisma/prisma#13967] [Q&A]
## Problem

<img width="660" alt="CleanShot 2022-06-21 at 22 15 42@2x" src="https://user-images.githubusercontent.com/284476/174928878-668f907e-ae48-4faa-9e02-c99c95e4fb77.png">

## Suggested solution

The only thing I'm allowed to do is add `@db.ObjectId`. Why can't the machine infer this from there being an `@id`. How is it helpful to have to manually add and generally read this attribute?

## Alternatives

N/A
## Additional context

N/A

**Comment by janpio** [comment]
Answer here is similar to https://github.com/prisma/prisma/discussions/13966#discussioncomment-3005070 - `@id` means one thing in PSL (PK, unique, model identifier), and the native type of a field is not one of the things it means - those are defined via `@db.ObjectId` attributes on the field.

**Reply by jasonkuhrt** [reply]
If the machine requires `@db.ObjectId` on every `@id` field then it has de-facto changed the meaning of `@id` in PSL through that effect, only worse so because its burdening the end user with the change.

I would be more understanding if there was some user choice here (but then I would advocate for better defaults), but there is not.

**Reply by janpio** [reply]
But then you are asking to redefine how `@id` works and what meaning it transports. By that argument you could also wrap in `@default(auto())` and the default field name `_id`. But that is just not how `@id` currently works.

**Reply by jasonkuhrt** [reply]
Do we agree that transports (meaning datasource providers) are currently redefining the requirements of `@id`? For example changing `mongodb` to `postgres` removes the requirement of using `@db.ObjectId`.

My claim is that a _constant requirement_ could be treated specially.

**Reply by tomhoule** [reply]
Let's recap from the more concrete points: the `_id` field on a mongo collection _can_ have another type, if you do not use `@default(auto())`. You can generate uuids application side and use that as `_id` fields in your collection, and both the database and prisma would be happy with that. `@default(auto())` means that we let the database generate `_id`, and it can only generate object ids.

So I'm not sure that was taken into account in your suggestion, but that only slightly changes the point: `@id` doesn't tell us what type the field is, but we could absolutely infer the native type from the combination of `@id` and `@default(auto())`.

I don't think we want to do that, because it doesn't make things simpler, it means more implicit behaviour that users need to understand to make sense of their schema, its behaviour in client and as a consequence the shape of their data. I think it would be hard for users to infer from `@default(auto())` that the values in that field will be kind of byte string (ObjectId) from the context they are given in the schema. If they want to _write_ to that field, we would have to refuse regular strings, or risk them mixing strings and objectids in the collection. Moreover, everywhere else in the schema, `myField String` means a regular string, only in this context it wouldn't. Any point about types and correctness is of course weaker on mongodb, but even in bson, objectid and string are different types. And mixing regular strings and object ids in the same collection for the same field has obvious bad consequences, so there are benefits to protecting users from that.

> Do we agree that transports (meaning datasource providers) are currently redefining the requirements of @id?

yep — especially mongo has very strict rules.

**Reply by jasonkuhrt** [reply]
Thanks @janpio and @tomhoule for the continued explanation.

> @default(auto()) means that we let the database generate _id, and it can only generate object ids.

This I did not know. So the "constant requirement" only kicks in given a _combination_. That at least makes me think twice right now.

> Moreover, everywhere else in the schema, myField String means a regular string, only in this context it wouldn't. Any point about types and correctness is of course weaker on mongodb, but even in bson, objectid and string are different types. And mixing regular strings and object ids in the same collection for the same field has obvious bad consequences, so there are benefits to protecting users from that.

So IIUC:

  1. Currently: `String` meaning is affected by presence of `@db.ObjectId`
  2. Why not: Just let `ObjectId` be put instead of `String`

As a user I can grok that the output of certain function accepted by `@default` places type constraints on the field its being used on. ✅ 

What is confusing to me is how indirectly this is currently achieved, by introducing a new syntactic place where field type is specified (from second column to trailing attribute). Not  only is the place changed though, but the old one `String` remains and it must be actively ignored IIUC.

**Reply by tomhoule** [reply]
> Why not: Just let ObjectId be put instead of String

and

> What is confusing to me is how indirectly this is currently achieved, by introducing a new syntactic place where field type is specified (from second column to trailing attribute)

is a state of things that has historical reasons that may or may not be good ones, depending on how you look at things.

The reason we have, for example `name String @db.VarChar(20)` instead of `name VarChar(20)` is that the client and QE do not care about native types, or at least they did not, originally. `String` is the client contract: this field accepts a string. But that is not enough information for accurate schema validation and migrations, there we need to know the exact type (`@db.VarChar(20)`) — that was the whole point of the scalar types initiative around the time of the Prisma 2 release (iirc).

Internally, we have a mapping for each connector from the native types to the smaller set of client/built-in scalar types, so if we designed things from scratch today, `name VarChar(20)` would be definitely possible (we can determine that the client should accept/return strings for that field), and possibly less confusing to users, but we landed on what we have now because it was the most practical/less breaking path at the time. My personal opinion on this is that given enough editor and generated types support, we might want to merge built-in scalar types and native types in the future, i.e. `id ObjectId`, but that's of course a relatively large breaking change.

**Reply by jasonkuhrt** [reply]
Thanks that's a helpful history lesson! Tension between intentional design and incidental history is always there and IIUC for Prisma the benefits that totally intentional design would bring still haven't outweighed the cost of deviating from the history part. Understandable, and like any investment its natural that either side of this ebb and flow!

I wouldn't say the net result of this thread leads me to a change in my experience/impression of PSL with Mongo, but it helps build empathy about _why_. I also may very well be in the niche minority category here of overall user impressions!

**Reply by janpio** [reply]
(We actually fully agree - being able to simplify this without breaking with the design and history would be great.)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
